numba/tests/test_np_functions.py

```diff
@@ -676,6 +700,63 @@ def test_diff2_exceptions(self):
                 cfunc(arr, n)
             self.assertIn("order must be non-negative", str(raises.exception))
 
+    def test_isobj_functions(self):
+        def values():
+            yield 1
+            yield 1 + 0j
+            yield np.asarray([3, 1 + 0j, True])
+
+        pyfuncs = [iscomplexobj, isrealobj]
+        for pyfunc in pyfuncs:
+            cfunc = jit(nopython=True)(pyfunc)
+            for x in values():
+                expected = pyfunc(x)
+                got = cfunc(x)
+                self.assertEqual(expected, got)
+
+    def test_is_real_or_complex(self):
+        def values():
+            yield np.array([1 + 1j, 1 + 0j, 4.5, 3, 2, 2j])
+            yield np.array([1, 2, 3])
+            yield 3
+            yield 1 + 4j
+            yield (1 + 4j, 2 + 0j)
+            yield np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
+
+        pyfuncs = [iscomplex, isreal]
+        for pyfunc in pyfuncs:
+            cfunc = jit(nopython=True)(pyfunc)
+            for x in values():
+                expected = pyfunc(x)
+                got = cfunc(x)
+                self.assertPreciseEqual(expected, got)
+
+    def test_isneg_or_ispos_inf(self):
+        def values():
+            yield np.NINF, None
+            yield np.inf, None
+            yield np.PINF, None
+            yield np.asarray([-np.inf, 0., np.inf]), None
```

Comment:
Seems like the resolution of https://github.com/numba/numba/pull/4610/files#r337510057 would suggest that `np.isneginf(np.NINF, np.empty(12))` would work too (1-to-many broadcast), perhaps add an example of that here?

Reply:
This input should also fail because of a broadcasting problem:
```
np.isneginf(np.array([np.NINF, np.NINF]), np.zeros((3,))) 
```
but currently does not.

Further, `np.isinf()` is going to need a minor update to deal with `array-like` things else most of these newly defined functions won't if e.g. `(np.inf, np.inf,)` is an arg (or restrictions can be applied in the above impls).

Reply:
Wasn't this a job for `np.logical_and` to check the arity of the arguments before applying the function?

Reply:
I think at run time the shapes need to be checked? The rule appears to be, if the input has `.ndim` of value `N` and `.size` of 1 and `N < out.ndim` then do broadcast `isneginf(input[0])->out`, else the standard shape matching rules apply i.e. shapes must be the same.

Seems like NumPy is using a `logical_and` because it is `AND`ing the "is it infinite" with an appropriate sign bit. I'm not sure how it's applicable to what you describe other than it conveniently errors out for things it cannot handle?

Reply:
Yeah, that seemed to solve the problem